src/neural/cuda/winograd_helper.inc

```diff
@@ -419,6 +419,24 @@ __device__ __forceinline__ float warpReduce(float x) {
   return x;
 }
 
+// fast max reduction for the warp
+__device__ __forceinline__ float warpMax(float x) {
+#pragma unroll
+  for (int mask = 16; mask > 0; mask >>= 1)
+    x = max(x, __shfl_xor_sync(0xFFFFFFFF, x, mask));
+
+  return x;
+}
+
+// atomic max implementation for floats
+__device__ __forceinline__ float atomicMaxFloat (float * addr, float val) {
+    float max;
+    max = (val >= 0) ? __int_as_float(atomicMax((int *)addr, __float_as_int(val))) :
```

Comment:
As mentioned in chat, stackoverflow suggests val >= 0 probably should be !signbit(val) to ensure a val of negative 0 doesn't get handled incorrectly as being smaller than existing negative values.